patch-package
--- a/node_modules/react-native-reanimated/src/core/AnimatedProps.js
+++ b/node_modules/react-native-reanimated/src/core/AnimatedProps.js
@@ -29,7 +29,7 @@ export function createOrReusePropsNode(props, callback, oldNode) {
     };
   }
   const config = sanitizeProps(props);
-  if (oldNode && deepEqual(config, oldNode._config)) {
+  if (oldNode) {
     return oldNode;
   }
   return new AnimatedProps(props, config, callback);
