diff --git a/node_modules/expo-device/ios/UIDevice.swift b/node_modules/expo-device/ios/UIDevice.swift
index 6866b89..b29f0a9 100644
--- a/node_modules/expo-device/ios/UIDevice.swift
+++ b/node_modules/expo-device/ios/UIDevice.swift
@@ -182,8 +182,13 @@ public extension UIDevice {
   // swiftlint:enable closure_body_length
 
   // Credit: https://github.com/developerinsider/isJailBroken/blob/master/IsJailBroken/Extension/UIDevice%2BJailBroken.swift
+
   var isSimulator: Bool {
-    return TARGET_OS_SIMULATOR != 0
+    #if targetEnvironment(simulator)
+      return true
+    #else
+      return false
+    #endif
   }
 
   var isJailbroken: Bool {
