diff --git a/node_modules/react-native-reanimated/Common/cpp/ReanimatedRuntime/RNRuntimeDecorator.cpp b/node_modules/react-native-reanimated/Common/cpp/ReanimatedRuntime/RNRuntimeDecorator.cpp
index c169431..d398502 100644
--- a/node_modules/react-native-reanimated/Common/cpp/ReanimatedRuntime/RNRuntimeDecorator.cpp
+++ b/node_modules/react-native-reanimated/Common/cpp/ReanimatedRuntime/RNRuntimeDecorator.cpp
@@ -38,7 +38,8 @@ void RNRuntimeDecorator::decorate(
   injectReanimatedCppVersion(rnRuntime);
 
   rnRuntime.global().setProperty(
-      rnRuntime, "_REANIMATED_IS_REDUCED_MOTION", isReducedMotion);
+      rnRuntime, "_REANIMATED_IS_REDUCED_MOTION", false // isReducedMotion
+      );
 
   rnRuntime.global().setProperty(
       rnRuntime,
diff --git a/node_modules/react-native-reanimated/android/src/main/cpp/NativeProxy.cpp b/node_modules/react-native-reanimated/android/src/main/cpp/NativeProxy.cpp
index fcd4a8a..de6e554 100644
--- a/node_modules/react-native-reanimated/android/src/main/cpp/NativeProxy.cpp
+++ b/node_modules/react-native-reanimated/android/src/main/cpp/NativeProxy.cpp
@@ -206,7 +206,7 @@ void NativeProxy::injectCppVersion() {
 void NativeProxy::installJSIBindings() {
   jsi::Runtime &rnRuntime = *rnRuntime_;
   WorkletRuntimeCollector::install(rnRuntime);
-  auto isReducedMotion = getIsReducedMotion();
+  auto isReducedMotion = false getIsReducedMotion();
   RNRuntimeDecorator::decorate(
       rnRuntime, nativeReanimatedModule_, isReducedMotion);
 #ifndef NDEBUG
@@ -232,8 +232,9 @@ void NativeProxy::performOperations() {
 }
 
 bool NativeProxy::getIsReducedMotion() {
-  static const auto method = getJniMethod<jboolean()>("getIsReducedMotion");
-  return method(javaPart_.get());
+  // static const auto method = getJniMethod<jboolean()>("getIsReducedMotion");
+  // return method(javaPart_.get());
+  return false
 }
 
 void NativeProxy::registerNatives() {
diff --git a/node_modules/react-native-reanimated/android/src/main/java/com/swmansion/reanimated/nativeProxy/NativeProxyCommon.java b/node_modules/react-native-reanimated/android/src/main/java/com/swmansion/reanimated/nativeProxy/NativeProxyCommon.java
index 427b6c8..03ee7e2 100644
--- a/node_modules/react-native-reanimated/android/src/main/java/com/swmansion/reanimated/nativeProxy/NativeProxyCommon.java
+++ b/node_modules/react-native-reanimated/android/src/main/java/com/swmansion/reanimated/nativeProxy/NativeProxyCommon.java
@@ -249,11 +249,12 @@ public abstract class NativeProxyCommon {
 
   @DoNotStrip
   public boolean getIsReducedMotion() {
-    ContentResolver mContentResolver = mContext.get().getContentResolver();
-    String rawValue =
-        Settings.Global.getString(mContentResolver, Settings.Global.TRANSITION_ANIMATION_SCALE);
-    float parsedValue = rawValue != null ? Float.parseFloat(rawValue) : 1f;
-    return parsedValue == 0f;
+    // ContentResolver mContentResolver = mContext.get().getContentResolver();
+    // String rawValue =
+    //     Settings.Global.getString(mContentResolver, Settings.Global.TRANSITION_ANIMATION_SCALE);
+    // float parsedValue = rawValue != null ? Float.parseFloat(rawValue) : 1f;
+    // return parsedValue == 0f;
+    return false
   }
 
   @DoNotStrip
diff --git a/node_modules/react-native-reanimated/apple/REAModule.mm b/node_modules/react-native-reanimated/apple/REAModule.mm
index 4148ea5..e7db97d 100644
--- a/node_modules/react-native-reanimated/apple/REAModule.mm
+++ b/node_modules/react-native-reanimated/apple/REAModule.mm
@@ -326,9 +326,9 @@ - (void)commonInit:(std::shared_ptr<NativeReanimatedModule>)nativeReanimatedModu
      withRnRuntime:(jsi::Runtime &)rnRuntime
 {
 #if __has_include(<UIKit/UIAccessibility.h>)
-  auto isReducedMotion = UIAccessibilityIsReduceMotionEnabled();
+  auto isReducedMotion = false // UIAccessibilityIsReduceMotionEnabled();
 #else
-  auto isReducedMotion = NSWorkspace.sharedWorkspace.accessibilityDisplayShouldReduceMotion;
+  auto isReducedMotion = false // NSWorkspace.sharedWorkspace.accessibilityDisplayShouldReduceMotion;
 #endif
   WorkletRuntimeCollector::install(rnRuntime);
   RNRuntimeDecorator::decorate(rnRuntime, nativeReanimatedModule, isReducedMotion);
diff --git a/node_modules/react-native-reanimated/lib/module/reanimated2/PlatformChecker.js b/node_modules/react-native-reanimated/lib/module/reanimated2/PlatformChecker.js
index 812f2e8..36d5419 100644
--- a/node_modules/react-native-reanimated/lib/module/reanimated2/PlatformChecker.js
+++ b/node_modules/react-native-reanimated/lib/module/reanimated2/PlatformChecker.js
@@ -36,8 +36,9 @@ export function isWindowAvailable() {
   return typeof window !== 'undefined';
 }
 export function isReducedMotion() {
-  return isWeb() ? isWindowAvailable() ?
-  // @ts-ignore Fallback if `window` is undefined.
-  !window.matchMedia('(prefers-reduced-motion: no-preference)').matches : false : !!global._REANIMATED_IS_REDUCED_MOTION;
+  // return isWeb() ? isWindowAvailable() ?
+  // // @ts-ignore Fallback if `window` is undefined.
+  // !window.matchMedia('(prefers-reduced-motion: no-preference)').matches : false : !!global._REANIMATED_IS_REDUCED_MOTION;
+  return false
 }
 //# sourceMappingURL=PlatformChecker.js.map
\ No newline at end of file
diff --git a/node_modules/react-native-reanimated/lib/module/reanimated2/hook/useReducedMotion.js b/node_modules/react-native-reanimated/lib/module/reanimated2/hook/useReducedMotion.js
index 92c571e..701f8ed 100644
--- a/node_modules/react-native-reanimated/lib/module/reanimated2/hook/useReducedMotion.js
+++ b/node_modules/react-native-reanimated/lib/module/reanimated2/hook/useReducedMotion.js
@@ -12,6 +12,7 @@ const IS_REDUCED_MOTION = isReducedMotion();
  * @see https://docs.swmansion.com/react-native-reanimated/docs/device/useReducedMotion
  */
 export function useReducedMotion() {
-  return IS_REDUCED_MOTION;
+  // return IS_REDUCED_MOTION;
+  return false
 }
 //# sourceMappingURL=useReducedMotion.js.map
\ No newline at end of file
diff --git a/node_modules/react-native-reanimated/src/reanimated2/PlatformChecker.ts b/node_modules/react-native-reanimated/src/reanimated2/PlatformChecker.ts
index bc22b42..b747e80 100644
--- a/node_modules/react-native-reanimated/src/reanimated2/PlatformChecker.ts
+++ b/node_modules/react-native-reanimated/src/reanimated2/PlatformChecker.ts
@@ -47,10 +47,11 @@ export function isWindowAvailable() {
 }
 
 export function isReducedMotion() {
-  return isWeb()
-    ? isWindowAvailable()
-      ? // @ts-ignore Fallback if `window` is undefined.
-        !window.matchMedia('(prefers-reduced-motion: no-preference)').matches
-      : false
-    : !!(global as localGlobal)._REANIMATED_IS_REDUCED_MOTION;
+  // return isWeb()
+  //   ? isWindowAvailable()
+  //     ? // @ts-ignore Fallback if `window` is undefined.
+  //       !window.matchMedia('(prefers-reduced-motion: no-preference)').matches
+  //     : false
+  //   : !!(global as localGlobal)._REANIMATED_IS_REDUCED_MOTION;
+  return false
 }
diff --git a/node_modules/react-native-reanimated/src/reanimated2/hook/useReducedMotion.ts b/node_modules/react-native-reanimated/src/reanimated2/hook/useReducedMotion.ts
index 076de31..018fa79 100644
--- a/node_modules/react-native-reanimated/src/reanimated2/hook/useReducedMotion.ts
+++ b/node_modules/react-native-reanimated/src/reanimated2/hook/useReducedMotion.ts
@@ -12,5 +12,6 @@ const IS_REDUCED_MOTION = isReducedMotion();
  * @see https://docs.swmansion.com/react-native-reanimated/docs/device/useReducedMotion
  */
 export function useReducedMotion() {
-  return IS_REDUCED_MOTION;
+  // return IS_REDUCED_MOTION;
+  return false
 }
